File: backend/src/routes/meals.js
const express = require('express');
const { body, validationResult } = require('express-validator');
const { authenticate, authorize } = require('../middleware/auth');
const db = require('../config/database');

const router = express.Router();
router.use(authenticate);

router.get('/', async (req, res, next) => {
  try {
    const { cafeteria_id, category } = req.query;
    let sql = 'SELECT * FROM meals WHERE is_active=true';
    const params = [];
    if (cafeteria_id) { params.push(cafeteria_id); sql += ` AND cafeteria_id=$${params.length}`; }
    if (category)     { params.push(category);     sql += ` AND category=$${params.length}`; }
    sql += ' ORDER BY name';
    const { rows } = await db.query(sql, params);
    res.json(rows);
  } catch (err) { next(err); }
});

router.post('/', authorize('admin', 'manager'), [
  body('name').trim().notEmpty(),
  body('category').isIn(['breakfast', 'lunch', 'dinner', 'snack']),
  body('cafeteria_id').isUUID(),
  body('serving_size_g').isInt({ min: 1 }),
  body('cost_per_serving').isFloat({ min: 0 }),
], async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) return res.status(400).json({ errors: errors.array() });

    const { name, category, cafeteria_id, serving_size_g, cost_per_serving, description } = req.body;
    const { rows } = await db.query(
      `INSERT INTO meals (name, category, cafeteria_id, serving_size_g, cost_per_serving, description)
       VALUES ($1,$2,$3,$4,$5,$6) RETURNING *`,
      [name, category, cafeteria_id, serving_size_g, cost_per_serving, description]
    );
    res.status(201).json(rows[0]);
  } catch (err) { next(err); }
});

router.get('/:id/waste-trend', async (req, res, next) => {
  try {
    const { rows } = await db.query(
      `SELECT DATE(created_at) AS date, SUM(weight_kg) AS total_waste_kg
       FROM waste_logs
       WHERE meal_id=$1 AND created_at >= NOW() - INTERVAL '14 days'
       GROUP BY DATE(created_at)
       ORDER BY date`,
      [req.params.id]
    );
    res.json(rows);
  } catch (err) { next(err); }
});

module.exports = router;

File: backend/routers/meals.py
from typing import Annotated
from uuid import UUID

import asyncpg
from fastapi import APIRouter, Depends

from core.database import get_conn
from core.security import get_current_user, require_roles
from models.schemas import MealCreate, MealOut

router = APIRouter(prefix="/meals", tags=["meals"])
Auth = Annotated[dict, Depends(get_current_user)]
Conn = Annotated[asyncpg.Connection, Depends(get_conn)]


@router.get("/", response_model=list[MealOut])
async def list_meals(user: Auth, conn: Conn, cafeteria_id: UUID | None = None, category: str | None = None):
    conditions = ["is_active=true"]
    params: list = []

    if cafeteria_id:
        params.append(cafeteria_id)
        conditions.append(f"cafeteria_id=${len(params)}")
    if category:
        params.append(category)
        conditions.append(f"category=${len(params)}")

    rows = await conn.fetch(
        f"SELECT * FROM meals WHERE {' AND '.join(conditions)} ORDER BY name",
        *params,
    )
    return [MealOut(**dict(r)) for r in rows]


@router.post("/", response_model=MealOut, status_code=201,
             dependencies=[Depends(require_roles("admin", "manager"))])
async def create_meal(body: MealCreate, conn: Conn):
    row = await conn.fetchrow(
        """INSERT INTO meals (name, category, cafeteria_id, serving_size_g, cost_per_serving, description)
           VALUES ($1,$2,$3,$4,$5,$6) RETURNING *""",
        body.name, body.category, body.cafeteria_id,
        body.serving_size_g, body.cost_per_serving, body.description,
    )
    return MealOut(**dict(row))


@router.get("/{meal_id}/waste-trend")
async def meal_waste_trend(meal_id: UUID, user: Auth, conn: Conn):
    rows = await conn.fetch(
        """SELECT DATE(created_at) AS date, SUM(weight_kg) AS total_waste_kg
           FROM waste_logs
           WHERE meal_id=$1 AND created_at >= NOW() - INTERVAL '14 days'
           GROUP BY DATE(created_at)
           ORDER BY date""",
        meal_id,
    )
    return [dict(r) for r in rows]
